L1Trigger/TrackerDTC/src/Setup.cc

```diff
@@ -476,13 +478,176 @@ namespace trackerDTC {
+        ptMin, ptMax, -etaMax, etaMax, tip, lip, minHit, signalOnly, intimeOnly, chargedOnly, stableOnly);
+    tpSelectorLoose_ = TrackingParticleSelector(ptMin, ptMax, -etaMax, etaMax, tip, lip, minHit, false, false, false, stableOnly);
+  }
+
+  //
+  int Setup::layerId(const TTStubRef& ttStubRef) const {
+    const DetId& detId = ttStubRef->getDetId();
+    return detId.subdetId() == StripSubdetector::TOB ? trackerTopology_->layer(detId)
+                                                     : trackerTopology_->tidWheel(detId) + offsetLayerDisks_;
+  }
+
+  //
+  int Setup::barrel(const TTStubRef& ttStubRef) const {
+    const DetId& detId = ttStubRef->getDetId();
+    return detId.subdetId() == StripSubdetector::TOB;
+  }
+
+  //
+  int Setup::psModule(const TTStubRef& ttStubRef) const {
+    const DetId& detId = ttStubRef->getDetId();
+    return trackerGeometry_->getDetectorType(detId) == TrackerGeometry::ModuleType::Ph2PSP;
+  }
+
+  //
+  TTBV Setup::layerMap(const vector<int>& ints) const {
+    TTBV ttBV;
+    for (int layer = numLayers_ - 1; layer >= 0; layer--) {
+      const int i = ints[layer];
+      ttBV += TTBV(i, kfWidthLayerCount_);
+    }
+    return ttBV;
+  }
+
+  //
+  TTBV Setup::layerMap(const TTBV& hitPattern, const vector<int>& ints) const {
+    TTBV ttBV;
+    for (int layer = numLayers_ - 1; layer >= 0; layer--) {
+      const int i = ints[layer];
+      ttBV += TTBV((hitPattern[layer] ? i - 1 : 0), kfWidthLayerCount_);
+    }
+    return ttBV;
+  }
+
+  //
+  vector<int> Setup::layerMap(const TTBV& hitPattern, const TTBV& ttBV) const {
+    TTBV bv(ttBV);
+    vector<int> ints(numLayers_, 0);
+    for (int layer = 0; layer < numLayers_; layer++) {
+      const int i = bv.extract(kfWidthLayerCount_);
+      ints[layer] = i + (hitPattern[layer] ? 1 : 0);
+    }
+    return ints;
+  }
+
+  //
+  vector<int> Setup::layerMap(const TTBV& ttBV) const {
+    TTBV bv(ttBV);
+    vector<int> ints(numLayers_, 0);
+    for (int layer = 0; layer < numLayers_; layer++)
+      ints[layer] = bv.extract(kfWidthLayerCount_);
+    return ints;
+  }
+
+  //
+  double Setup::dZ(const TTStubRef& ttStubRef, double cot) const {
+    const DetId& detId = ttStubRef->getDetId();
+    SensorModule* sm = sensorModule(detId + 1);
+    const double sigma = sm->pitchCol() * sm->tiltCorrection(cot);
+    const double digi = tmttBaseZ_;
+    const double dZ = sigma + digi;
+    if (dZ >= maxdZ_ || dZ < mindZ_) {
+      cms::Exception exception("out_of_range");
+      exception.addContext("trackerDTC::Setup::dZ");
+      exception << "Stub z uncertainty " << dZ << " "
+                << "is out of range " << mindZ_ << " to " << maxdZ_ << ".";
+      throw exception;
+    }
+    return dZ;
+  }
+
+  //
+  double Setup::v1(const TTStubRef& ttStubRef, double cot) const {
+    const DetId& detId = ttStubRef->getDetId();
+    SensorModule* sm = sensorModule(detId + 1);
+    const double sigma = pow(sm->pitchCol() * sm->tiltCorrection(cot), 2) / 12.;
```

Comment:
This eqn. is inconsistent with that in dZ(), since this one divides sigma^2 by 12. Which is best? Logically, both the pitch & digi granularity should be divided by sqrt(12).

Reply:
added / 12 for v0 and v1 digi component.

Reply:
This means that dPhi() & dZ() represent the nominal uncertainty multiplied by sqrt(12), (if one considers only the pitch & digi). This should be mentioned in the comment for them in Setup.h. However, the treatment of scattering in v0() & dPhi() is not consistent with this sqrt(12) scaling, since both use identical scattering uncertainty.

Reply:
you are sayng that the e.g. the number 0.00075 is wrong by sqrt(12)?